smolyay/surrogate.py

```diff
@@ -0,0 +1,279 @@
+
+        return self._coefficients.tolist()
+
+    def train_from_data(self, real_function_at_grids, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        real_function_at_grids: list
+            Real function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        """
+        if len(real_function_at_grids) != len(self._grids.points):
+            raise IndexError("Data must be in length of the grid points.")
+        self.real_function_at_grids = real_function_at_grids
+        self.linear_solver = linear_solver
+        self._make_basis_matrix()
+        self._coefficients = solve_linear(self._basis_matrix,
+                                          self.real_function_at_grids,
+                                          self.linear_solver)
+
+        return self._coefficients.tolist()
+
+    def __call__(self, x):
+        """Evaluate surrogate at a given input.
+
+        Parameter
+        ---------
+        x: list (for dimension > 1), or float (for dimension = 1)
+            Input.
+
+        Returns
+        -------
+        float
+            Surrogate output at x.
+
+        Raises
+        ------
+        ValueError
+            For surrogate to be evaluated, function needs to be trained.
+
+        IndexError
+            Input of the surrogate must be of length of the
+            real function's dimensionality.
+        """
+        if self._coefficients is None:
+            raise ValueError('Function needs training!')
+        if isinstance(x, (int, float)):
+            x = [x]
+        if len(x) != self._dimension:
+            raise IndexError('Input must be of length of the dimension.')
+
+        # transform grids into basis domain
+        input_surrogate = []
+        for dimension_ in range(self._dimension):
+            basis_domain_grid = numpy.polynomial.polyutils.mapdomain(
+                x[dimension_], self.domain[dimension_], (-1, 1))
+            input_surrogate.append(basis_domain_grid)
+
+        # evaluate surrogate
+        surrogate_output = 0
+        for index_point in range(len(self._grids.points)):
+            output = 1
+            for dimension_ in range(self._dimension):
+                output *= (self._grids.basis_functions[index_point]
+                           [dimension_](input_surrogate[dimension_]))
+            output *= self._coefficients[index_point]
+            surrogate_output += output
+
+        return surrogate_output
+
+
+def solve_linear(a, b, method):
```

Comment:
If you can make `train()` use `train_from_data()`, then these if statements could just be put in `train_from_data` rather than as a separate function.